### PR TITLE
Clean-up and refactor of attribute and attribute tags classes; prepar…

### PR DESCRIPTION
**Pull request by bmaizels**
…ation for injection of custom converters

## Description
Besides the clean-up, one key structural feature of this change is to delay the resolution of attributes
to specific converters so that a custom converter provider can be injected. The interface changes required to facilitate the actual injection are coming in a separate PR.

Relates to https://github.com/aws/aws-sdk-java-v2/issues/35

## License

- [X] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/mapper/StaticTableMetadata.java [hunk 1]
```diff
@@ -166,37 +171,31 @@ public int hashCode() {
...
 
-    @ThreadSafe
+    /**
+     * Builder for {@link StaticTableMetadata}
+     */
     public static class Builder {
-        private final Map<String, Object> customMetadata = new ConcurrentHashMap<>();
-        private final Map<String, Index> indexByNameMap = new ConcurrentHashMap<>();
-        private final Map<String, AttributeValueType> keyAttributes = new ConcurrentHashMap<>();
+        private final Map<String, Object> customMetadata = new HashMap<>();
```
Comment (cenedhryn) [inline services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/mapper/StaticTableMetadata.java]: Why did you change to HashMap?

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/AsyncBasicCrudTest.java [hunk 2]
```diff
@@ -170,27 +168,43 @@ public int hashCode() {
...
+                                                           .setter(Record::setId)
+                                                           .tags(primaryPartitionKey()))
+                         .addAttribute(String.class, a -> a.name("sort")
+                                                           .getter(Record::getSort)
+                                                           .setter(Record::setSort)
+                                                           .tags(primarySortKey()))
+                         .addAttribute(String.class, a -> a.name("attribute")
+                                                           .getter(Record::getAttribute)
+                                                           .setter(Record::setAttribute))
+                         .addAttribute(String.class, a -> a.name("attribute2*")
```
Comment (cenedhryn) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/AsyncBasicCrudTest.java]: I have noticed that in many test files, we're defining but not using the secondary index. Should we consider adding more tests to target that and cover more behavior?

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/VersionedRecordExtension.java [hunk 3]
```diff
@@ -21,15 +21,16 @@
...
+import software.amazon.awssdk.enhanced.dynamodb.AttributeValueType;
 import software.amazon.awssdk.enhanced.dynamodb.DynamoDbEnhancedClientExtension;
 import software.amazon.awssdk.enhanced.dynamodb.Expression;
 import software.amazon.awssdk.enhanced.dynamodb.TableMetadata;
 import software.amazon.awssdk.enhanced.dynamodb.internal.operations.OperationContext;
-import software.amazon.awssdk.enhanced.dynamodb.mapper.AttributeTag;
-import software.amazon.awssdk.enhanced.dynamodb.mapper.AttributeValueType;
+import software.amazon.awssdk.enhanced.dynamodb.mapper.StaticAttributeTag;
+import software.amazon.awssdk.enhanced.dynamodb.mapper.StaticTableMetadata;
 import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
```
[the remark this answers is not in the document]
Comment (bmaizels) [reply services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/VersionedRecordExtension.java]: Will fix.

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/converters/attribute/OptionalAttributeConvertersTest.java [hunk 4]
```diff
@@ -16,10 +16,10 @@
 package software.amazon.awssdk.enhanced.dynamodb.converters.attribute;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static software.amazon.awssdk.enhanced.dynamodb.internal.converter.attribute.EnhancedAttributeValue.fromNumber;
-import static software.amazon.awssdk.enhanced.dynamodb.internal.converter.attribute.EnhancedAttributeValue.nullValue;
 import static software.amazon.awssdk.enhanced.dynamodb.converters.attribute.ConverterTestUtils.transformFrom;
 import static software.amazon.awssdk.enhanced.dynamodb.converters.attribute.ConverterTestUtils.transformTo;
+import static software.amazon.awssdk.enhanced.dynamodb.internal.converter.attribute.EnhancedAttributeValue.fromNumber;
```
[the remark this answers is not in the document]
Comment (bmaizels) [reply services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/converters/attribute/OptionalAttributeConvertersTest.java]: Yes, I applied a consistent set of import rules across the whole project as I was already touching >50% of the files in this PR.

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/mapper/StaticAttributeTag.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.enhanced.dynamodb.AttributeValueType;
+
+/**
+ * Interface for a tag that can be applied to any {@link StaticAttribute}. When a tagged attribute is added to a
+ * {@link software.amazon.awssdk.enhanced.dynamodb.TableSchema}, the table metadata stored on the schema will be updated
+ * by calling the {@link #modifyMetadata(String, AttributeValueType)} method for every tag associated with the
+ * attribute.
+ * <p>
+ * Core implementations of these tags that
```
[the remark this answers is not in the document]
Comment (bmaizels) [reply services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/mapper/StaticAttributeTag.java]: Will fix.

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/mapper/StaticTableMetadata.java (hunk 1 above)
Comment (bmaizels) [reply services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/mapper/StaticTableMetadata.java]: I realized there was no value in maintaining this class as threadsafe as it was an immutable class and we can just follow our standard patterns.

Comment (bmaizels) [reply services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/mapper/StaticTableMetadata.java]: Well to be more specific we don't require any of our builders to be threadsafe. It's generally well understood they are not.

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/AsyncBasicCrudTest.java (hunk 2 above)
Comment (bmaizels) [reply services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/AsyncBasicCrudTest.java]: Query and scan are the only two operations currently supported by secondary indices and we have good test coverage on those. These tests share a lot of the POJOs so in a way what is being tested here is that non-index operations work fine when indexes are present.

Comment (cenedhryn) [reply services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/AsyncBasicCrudTest.java]: ok

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/mapper/StaticAttributeTag.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bmaizels) [reply services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/mapper/StaticAttributeTag.java]: Oddly enough I can't find it, I wonder if it got crunched in a merge. We'll see what happens when I push.

[the remark this answers is not in the document]
Comment (bmaizels) [reply services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/mapper/StaticAttributeTag.java]: Oh never mind I was looking at table tags, not attribute tags. Fixed.
